src/rust/engine/src/core.rs

```diff
@@ -287,6 +287,12 @@ pub enum Failure {
   /// A Node failed because a filesystem change invalidated it or its inputs.
   /// A root requestor should usually immediately retry their request.
   Invalidated,
+  /// A uncacheable node exhausted retry attempts for its dependencies because
+  /// they were invalidated too many times while running. This enum variant is distinguished
+  /// from Invalidated because we DONT want the scheduler to retry the request. Uncacheable
+  /// nodes are allowed to have side-effects and they shouldn't be run multiple times in one
+  /// session.
+  Exhausted,
```

Comment:
Rather than adding a new type to this enum, it should be fine to use `Throw(...)` here via https://github.com/pantsbuild/pants/blob/859bd107d958daf915d46277d71c6b6ebb9195e8/src/rust/engine/src/nodes.rs#L68-L74 ... possibly ditto the `FileWatch` exception, but not a big deal.

Adding the method to the `Node` trait is fine though... it just doesn't need to return a unique type.

Reply:
Gotcha, I assumed Throw was reserved for actual exceptions raised in python land.

Reply:
Done, for both Exhausted and FIleWatch.